watchmal/engine/reconstruction.py

```diff
@@ -0,0 +1,422 @@
+"""
+Class for training a fully supervised classifier
+"""
+
+# generic imports
+import numpy as np
+from datetime import timedelta
+from datetime import datetime
+from abc import ABC, abstractmethod
+import logging
+
+# hydra imports
+from hydra.utils import instantiate
+
+# torch imports
+import torch
+from torch.nn.parallel import DistributedDataParallel
+
+# WatChMaL imports
+from watchmal.dataset.data_utils import get_data_loader
+from watchmal.utils.logging_utils import CSVLog
+
+log = logging.getLogger(__name__)
+
+
+class ReconstructionEngine(ABC):
+    def __init__(self, truth_key, model, rank, gpu, dump_path):
+        """
+        Parameters
+        ==========
+        truth_key : string
+            Name of the key for the target values in the dictionary returned by the dataloader
+        model
+            `nn.module` object that contains the full network that the engine will use in training or evaluation.
+        rank : int
+            The rank of process among all spawned processes (in multiprocessing mode).
+        gpu : int
+            The gpu that this process is running on.
+        dump_path : string
+            The path to store outputs in.
+        """
+        # create the directory for saving the log and dump files
+        self.epoch = 0
+        self.step = 0
+        self.iteration = 0
+        self.best_validation_loss = 1.0e10
+        self.dump_path = dump_path
+        self.rank = rank
+        self.model = model
+        self.device = torch.device(gpu)
```

Comment:
Can this also be a CPU? As changed in the main function?

Reply:
Yeah, thanks for spotting that. We could rename this parameter to `device` instead of `gpu` because it does actually work with CPU instead.

Looks like in main.py [here](https://github.com/WatChMaL/WatChMaL/pull/31/files/2ad1c240f5d184e947b361416e9a950a33a5d89d#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R109) it was renamed from `gpu` to `device` for that reason, so I can change it in the engine too for consistency.

Reply:
just pushed a commit that updated it from `gpu` to `device`